File: accepted/300-longest-increasing-subsequence.c
/* https://leetcode.com/problems/longest-increasing-subsequence/ */

int lengthOfLIS(int* nums, int numsSize) 
{
    if (0 == numsSize)
        return 0;
    
    int nRet = 1;
    int *dp = (int *)malloc(numsSize * sizeof(int));
    dp[0] = 1;
    
    for (int i = 1; i < numsSize; i++)
    {
        int nResult = 1;
        for (int j = 0; j < i; j++)
            if (dp[j] >= nResult && nums[i] > nums[j])
                nResult = dp[j] + 1;
        
        dp[i] = nResult;
        nRet = nResult > nRet ? nResult : nRet;
    }
    free((void *)dp);
    
    return nRet;
}

File: accepted/300-longest-increasing-subsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        size = len(nums)
        if 0 == size:
            return 0
        
        dp = [0] * size
        dp[0] = 1
        
        for i in xrange(1, size):
            result = 1
            for j in xrange(i):
                if dp[j] >= result and nums[i] > nums[j]:
                    result = dp[j] + 1
            
            dp[i] = result
        
        return max(dp)
